katdal/applycal.py

```diff
@@ -216,57 +219,88 @@ def calc_gain_correction(sensor, index):
     return np.reciprocal(smooth_gains)
 
 
-def add_applycal_sensors(cache, attrs, data_freqs):
-    """Add virtual sensors that store calibration corrections, to sensor cache.
+def add_applycal_sensors(cache, attrs, data_freqs, cal_stream='cal',
+                         cal_substreams=None):
+    """Register virtual sensors for one calibration stream.
+
+    This operates on a single calibration stream called `cal_stream` (possibly
+    an alias), which derives from one or more underlying cal streams listed in
+    `cal_substreams` and has stream attributes in `attrs`.
+
+    The first set of virtual sensors maps all cal products into a unified
+    namespace (template 'Calibration/Products/`cal_stream`/{product_type}').
+    Map receptor inputs to the relevant indices in each calibration product
+    based on the ants and pols found in `attrs`. Then register a virtual sensor
+    per product type and per input in the SensorCache `cache`, with template
+    'Calibration/Corrections/`cal_stream`/{product_type}/{inp}'. The virtual
+    sensor function picks the appropriate correction calculator based on the
+    cal product type, which also uses auxiliary info like the channel
+    frequencies, `data_freqs`.
 
-    This maps receptor inputs to the relevant indices in each calibration
-    product based on the ants and pols found in `attrs`. It then registers
-    a virtual sensor per input and per cal product in the SensorCache `cache`,
-    with template 'Calibration/{inp}_correction_{product}'. The virtual sensor
-    function picks the appropriate correction calculator based on the cal
-    product name, which also uses auxiliary info like the channel frequencies,
-    `data_freqs`.
+    Parameters
+    ----------
+    cache : :class:`~katdal.sensordata.SensorCache` object
+        Sensor cache serving cal product sensors and receiving correction sensors
+    attrs : dict-like
+        Calibration stream attributes (e.g. a "cal" telstate view)
+    data_freqs : array of float, shape (*F*,)
+        Centre frequency of each frequency channel of visibilities, in Hz
+    cal_stream : string, optional
+        Name of (possibly virtual) calibration stream (e.g. "L1")
+    cal_substreams : sequence of string, optional
+        Names of actual underlying calibration streams (e.g. ["cal"]),
+        defaults to [`cal_stream`] itself
```

Comment:
It looks like it currently only uses the first element. Is this one of those places that you expect to extend for L2?

Reply:
Yes.